### ISSUE TITLE
Introduce support for parameterized tests

### ISSUE THREAD

**Issue by elygre** [junit-team/junit5#14]
## Deliverables

- [x] Introduce `@TestTemplate` and accompanying extension point.
- [x] Determine if #704 should be included in M4.
- [x] Determine if we need an `@InvocationIndex` annotation or `TestTemplateInfo` type for injecting the current _invocation index_, etc. into `@Test`, `@BeforeEach`, and `@AfterEach` methods.
  - See https://github.com/junit-team/junit5/commit/d2e3a4dfdc8cc4b4b8609e4297f8e00b753cb5b3
- [x] Introduce `junit-jupiter-params` module
  - [x] Merge initial contribution (#728)
  - [x] Javadoc
    - [x] `@since` tags
  - [x] API annotations
  - [x] Add unit tests where missing (see TODOs)
  - [x] Resolve TODOs
- [X] Document `@TestTemplate` and accompanying extension point ~~in User Guide and~~ Release Notes
  - [X] ~~Provide example(s) in the `documentation` module and include in the User Guide.~~ Will be addresses in #762
- [x] Document parameterized tests in User Guide and Release Notes
  - [x] Document the `junit-jupiter-params` module in the _Dependency Metadata_ section.
  - [x] Document the `junit-jupiter-params` module in the _Dependency Diagram_.
  - [x] Provide example(s) in the `documentation` module and include in the User Guide.

----

## Original Issue Description

At the time of writing, the https://github.com/junit-team/junit-lambda/wiki/Prototype-Test-Decorators page does not have all that much detail, so this might already be considered.

It would be good if the library shipped with ready-to-use simple parameter annotations that could be used inside a single test instance. It would in many ways look like JUnit 4's `@Parameterized` Runner, but for use inside a single class. When running the test, each `@Parameterized` method would appear many times, once for each data set.

``` java
class MyTest {

   // A simple test with a specified data set
   @Test @Parameterized("mod10")
   void testMod10(String value) { ... }

   @Parameterized("mod10") 
   Collection<Object[]> getMod10Values { ... }

   // Also including a method naming convention
   @Test @Parameterized("mod11", name="{0}")
   void testMod11(String value) { ... }

   @Parameterized("mod11")
   Collection<Object[]> getMod11Values { ... }

   // In the absense of a @Parameterized data provider, use reflection
   @Test @Parameterized("getMod12Values", name="{0}")
   void testMod12(String value) { ... }

   Collection<Object[]> getMod12Values{ ... }
}
```

----

## Related Issues

- #378
- #386 
- #604 
- #704 
- #728

**Comment by sbrannen** [comment]
@elygre, thanks for raising this issue.

We will certainly provide a mechanism for _parameterized_ tests. We just haven't gotten to it yet in the prototype.

However, we do already have generic support for supplying parameters to methods. This can be achieved by implementing a custom [`MethodParameterResolver`](https://github.com/junit-team/junit-lambda/blob/master/junit5-api/src/main/java/org/junit/gen5/api/extension/MethodParameterResolver.java).

Thus, the `MethodParameterResolver` support is the first building block required for _parameterized test_ support. The next step will be to provide test decorators a mechanism for _registering_ tests dynamically.

We will post back here once we begin work on that front.

Regards,

Sam

**Comment by marcphilipp** [comment]
_in progress_

**Comment by sbrannen** [comment]
@marcphilipp, whenever you start active work on this feature, let's make sure we introduce abstractions that are usable across numerous unsupported use cases (i.e., features in the backlog like this one, support for executing dynamic tests with parameters, etc.).

With that in mind, a few ideas I've been tossing around in my head are `TestOrchestrator` and `ExecutableTest`. 

`TestOrchestrator` would be an `Extension` API for extensions that are responsible for _orchestrating_ execution of executable tests (i.e., instances of `ExecutableTest`). This would include ordering (e.g., for scenario tests), threading, class loaders, etc.

An `ExecutableTest` would encapsulate a _test_ such as an `@Test` method, a `DynamicTest` generated by an `@TestFactory` method, a parameterized test method, etc. The idea here is to encapsulate the execution details of executable tests while allowing them to otherwise be treated the same in terms of orchestration.

Food for thought!

Sam

**Comment by nicolaiparlog** [comment]
Should there be a meta issue that collects the "numerous unsupported use cases" and allows for discussions of proposals? It would make it easier to see the scope of what is tackled and compare proposals.

I would like to chime in with some thoughts I had regarding lifecycle integration of dynamic tests and their factory methods.

**Comment by marcphilipp** [comment]
> Should there be a meta issue that collects the "numerous unsupported use cases" and allows for discussions of proposals?

Is there a way to provide a bit more restrictions on the topic? Otherwise, this might get out of hand quickly. 🤔

> I would like to chime in with some thoughts I had regarding lifecycle integration of dynamic tests and their factory methods.

Please share them in #378. 🙂

**Comment by elygre** [comment]
As time passes, I've come to realize that our most frequent use case has a few simple alternative values. Hence, I'd like to suggest an optional simplification:

```
   // A simple test with a hard coded data set
   @Test @Parameterized(values={ "github", "gitlab" })
   void testProviders (String providerName) { ... }
```

**Comment by marcphilipp** [comment]
I've updated the issue description and added deliverables.

**Comment by marcphilipp** [comment]
Initial support for parameterized tests is now in master (#728).

@junit-team/junit-lambda I've updated the deliverables of this task to reflect the remaining work that needs to be done. I'm un-assigning myself from this task because I will only have very little time in the upcoming three weeks.

**Comment by sbrannen** [comment]
FYI: I added further concrete deliverables regarding examples in the `documentation` project and User Guide.

**Comment by sbrannen** [comment]
@marcphilipp, why is a `TestTemplateInvocationContextProvider` required to return a non-empty `Stream` in the `provide()` method?

What if the extension determines only while _attempting to provide contexts_ that there are currently no contexts (for whatever reason)?

Playing devil's advocate: the current JavaDoc is only partially true. If a only a single provider is registered and it returns an empty stream, then the requirement holds true. However, if multiple providers are registered and at least one of them provides a context, then the other providers are free to return empty streams and no exception will be thrown.

In other words, we are not actually enforcing the document requirement at the moment. (at least that's my interpretation of the code in `master`)

Thoughts?

**Comment by marcphilipp** [comment]
Point taken. If we allow empty streams, I guess we could remove `supports` altogether. 

I'll start to work on the Javadoc while traveling tomorrow.

**Comment by sbrannen** [comment]
I think it's probably a good idea to keep `supports` around, even if technically only as an indicator or a defensive measure.

I imagine there will be authors of such extensions that _provide_ something even if it's not _supported_ (i.e., they will blindly generate something), but perhaps I'm overthinking things. 😉

**Comment by sbrannen** [comment]
FYI: I added a new sub-task to the _Deliverables_ regarding an `@InvocationIndex` annotation for injecting the current _invocation index_ into `@Test`, `@BeforeEach`, and `@AfterEach` methods.

**Comment by sbrannen** [comment]
If you look at the implementation of the `@InvocationIndex` support you will notice that there is a disconnect between the `InvocationIndexParameterResolver` and `TestTemplateInvocationContext` with regard to the _current invocation index_.

Specifically, `RepeatedTestInvocationContext` accepts the index via its constructor and via the `getDisplayName()` method. Ideally, we should make the _current invocation index_ available via a different (consistent) mechanism internally, but since I'm working on a branch devoted to repeated tests, I didn't want to start overhauling the _test template_ infrastructure.

@marcphilipp, as the mastermind behind test templates, do you have any thoughts on how you'd like to revise this?

**Comment by sbrannen** [comment]
Following up on my previous comment...

One option would be to introduce `int invocationIndex` as an argument to `TestTemplateInvocationContext.getAdditionalExtensions()`.

**Comment by sbrannen** [comment]
On second thought, `RepeatedTestInvocationContext` deals with the _repetition count_ and not necessarily with the _invocation index_. They are likely the same value, but if multiple `TestTemplateInvocationContextProviders` are registered then the two values will potentially differ (depending on the order in which the providers are registered).

So, in that regard, there are two concepts at play here:

1. repetition count for repeated tests
2. invocation index for any test template

When I spell it out like that, it strikes me that the generic _invocation index_ may actually not be all that useful; whereas, the repetition count is certainly useful.

And with that in mind, it might be better to drop the generic `@InvocationIndex` support and replace it with something like `@Repetition`.

@junit-team/junit-lambda, Thoughts?

**Comment by sbrannen** [comment]
@marcphilipp, can you shed some light on possible use cases for registering multiple `TestTemplateInvocationContextProviders` for the same test template method?

**Comment by sbrannen** [comment]
**Update**: I have opted to remove support for `@InvocationIndex` in the branch for repeated tests. As an alternative I have introduced `RepetitionInfo` which can be injected into methods, analogous to `TestInfo`. Details here: https://github.com/junit-team/junit5/commit/6f615a83a94a4ed37a91932e2f371a1cea2f9195

**Comment by sbrannen** [comment]
@marcphilipp, to answer my previous question on my own...

I discovered this example:

```java
@ParameterizedTest
@CsvSource("foo")
@MethodSource(names = { "first", "second" })
@ArgumentsSource(MyArgumentsProvider.class)
@CsvFileSource(resources = { "/single-column.csv", "two-column.csv" })
void testWithMultipleDifferentSources(String parameter) {
}
```

So that obviously works. 😉 

I don't think I would personally use that feature (combining multiple sources for the same parameterized test), but I suppose some people will find it useful.

Out of curiosity, is this use case something you've seen in the wild or had a feature request for?

**Comment by sbrannen** [comment]
@junit-team/junit-lambda, we need to decide if it makes sense to have something like `@InvocationIndex` or to just scrap the idea.

**Comment by sbrannen** [comment]
FYI: #704 is a follow-up to this issue.

**Comment by sbrannen** [comment]
FYI: as a follow up to the discussion regarding whether or not we need `supports()` for providers, the answer is a definitive "yes". Otherwise, the support for `@RepeatedTest` (and any other use cases where a provider is _registered_ but not used for a particular test template method) will be broken.

**Comment by sbrannen** [comment]
Regarding the need for an `@InvocationIndex` annotation, I have personally decided against it. However, we could provide an alternative means to access similar information about the current test template invocation -- something like `TestTemplateInfo` perhaps.

See comments here: https://github.com/junit-team/junit5/issues/704#issuecomment-289224642

**Comment by marcphilipp** [comment]
Thanks for all the hard work, guys!

Remaining work (`@TestTemplate` documentation) will be addresses in #762.
